docs/question/server.md

```diff
@@ -107,18 +209,31 @@ The `pl.to_json` function supports keyword-only options for different types of e
 From within `server.py` functions, directories can be accessed as:
 
 ```python
-data["options"]["question_path"]                      # on-disk location of the current question directory
-data["options"]["client_files_question_path"]         # on-disk location of clientFilesQuestion/
-data["options"]["client_files_question_url"]          # URL location of clientFilesQuestion/ (only in render() function)
-data["options"]["client_files_question_dynamic_url"]  # URL location of dynamically-generated question files (only in render() function)
-data["options"]["client_files_course_path"]           # on-disk location of clientFilesCourse/
-data["options"]["client_files_course_url"]            # URL location of clientFilesCourse/ (only in render() function)
-data["options"]["server_files_course_path"]           # on-disk location of serverFilesCourse/
+# on-disk location of the current question directory
+data["options"]["question_path"]
+
+# on-disk location of clientFilesQuestion/
+data["options"]["client_files_question_path"]
+
+# URL location of clientFilesQuestion/ (only in render() function)
+data["options"]["client_files_question_url"]
+
+# URL location of dynamically-generated question files (only in render() function)
+data["options"]["client_files_question_dynamic_url"]
+
+# on-disk location of clientFilesCourse/
+data["options"]["client_files_course_path"]
+
+# URL location of clientFilesCourse/ (only in render() function)
+data["options"]["client_files_course_url"]
+
+# on-disk location of serverFilesCourse/
+data["options"]["server_files_course_path"]
 ```
 
-## Generating dynamic files
+## Generating dynamic files with `file()`
 
-You can dynamically generate file objects in `server.py`. These files never appear physically on the disk. They are generated in `file()` and returned as strings, bytes-like objects, or file-like objects. A complete `question.html` and `server.py` example using a dynamically generated `fig.png` looks like:
+You can dynamically generate file objects in `server.py`. These files never appear physically on the disk. They are generated in `file()` and returned as strings, bytes-like objects, or file-like objects. These files are generated before the question is rendered in `question.html`. A complete `question.html` and `server.py` example using a dynamically generated `fig.png` looks like:
```

Comment:
> These files are generated before the question is rendered in `question.html`.

This is not true; did you get this impression from somewhere in the docs? If so, let's fix wherever that was to clear things up.

Reply:
This was an incorrect guess; can you suggest the correct text?

Reply:
This will help me inform https://github.com/PrairieLearn/PrairieLearn/pull/11830#discussion_r2054537142

Reply:
It needs to happen before the page is actually shown to the user, so I guess it happens after `render` but before it's actually rendered up? Or is this `file` function called dynamically upon requests?

Reply:
> Or is this `file` function called dynamically upon HTML requests?

This is exactly correct. Breadcrumbs:

- https://github.com/PrairieLearn/PrairieLearn/blob/396f5428db5d517b7e37ae47dd021553439cbde5/apps/prairielearn/src/question-servers/freeform.js#L1233
- https://github.com/PrairieLearn/PrairieLearn/blob/330f53fb3ea27caf759e6a2ace840087ec7760a6/apps/prairielearn/src/lib/question-render.ts#L222
- https://github.com/PrairieLearn/PrairieLearn/blob/330f53fb3ea27caf759e6a2ace840087ec7760a6/apps/prairielearn/src/pages/generatedFilesQuestion/generatedFilesQuestion.ts#L44-L51

Reply:
Updated in [9b81bf8](https://github.com/PrairieLearn/PrairieLearn/pull/11830/commits/9b81bf837e7c933a426590a9586f3c3b9bf27c5b).